src/core/thread/link_quality.hpp

```diff
@@ -50,6 +50,68 @@ namespace ot {
  * @{
  */
 
+/**
+ * This class implement an operation Success Rate Tracker.
+ *
+ * This can be used to track different link quality related metrics, e.g., CCA failure rate, frame tx success rate).
+ * The success rate is maintained using an exponential moving IIR averaging filter with a `uint16_t` as the storage.
+ *
+ */
+class SuccessRateTracker
+{
+public:
+    enum
+    {
+        kMaxRateValue = 0xffff,    ///< Indicates value corresponding to maximum (failure/success) rate of 100%.
+    };
+
+    /**
+     * This constructor initializes a `SuccessRateTracker` instance.
+     *
+     * After initialization the tracker starts with success rate 100% (failure rate 0%).
+     *
+     */
+    SuccessRateTracker(void): mFailureRate(0) { }
+
+    /**
+     * This method resets the tracker to its initialized state, setting success rate to 100%.
+     *
+     */
+    void Reset(void) { mFailureRate = 0; }
+
+    /**
+     * This method adds a sample (success or failure) to `SuccessRateTracker`.
+     *
+     * @param[in] aSuccess   The sample status be added, `true` for success, `false` for failure.
+     * @param[in] aWeight    The weigh coefficient used for adding the new sample into average.
+     *
+     */
+    void AddSample(bool aSuccess, uint16_t aWeight = kDefaultWeight);
+
+    /**
+     * This method returns the average failure rate.
+     *
+     * @retval the average failure rate `[0-kMaxRateValue]` with `kMaxRateValue` corresponding to 100%.
+     *
+     */
+    uint16_t GetFailureRate(void) const { return mFailureRate; }
+
+    /**
+     * This method returns the average success rate.
+     *
+     * @retval the average success rate as [0-kMaxRateValue] with `kMaxRateValue` corresponding to 100%.
+     *
+     */
+    uint16_t GetSuccessRate(void) const { return kMaxRateValue - GetFailureRate(); }
+
+private:
+    enum
+    {
+        kDefaultWeight = 64,
+    };
+
+    uint16_t mFailureRate;
```

Comment:
Any reason why the `SuccessRateTracker` actually keeps track of failure rate instead?

Reply:
Not really! I always worked out the math considering average failure rate being maintained. 
Originally called the class `RateTracker` but then changed it to `SuccessRateTracker` to be a bit more descriptive. Can change this.